File: algorithms/medium/short_encoding_of_words.ex
defmodule Solution do
  defmodule Trie do
    defstruct children: %{}

    def new(), do: %Trie{}

    def insert(nil, cs), do: insert(Trie.new(), cs)

    def insert(tree, []), do: tree

    def insert(tree, [c | cs]) do
      %{tree | children: Map.put(tree.children, c, insert(tree.children[c], cs))}
    end

    def dfs(tree, len, sum) do
      if map_size(tree.children) == 0 do
        sum + len + 1
      else
        tree.children
        |> Map.values()
        |> Enum.reduce(sum, fn v, acc ->
          dfs(v, len + 1, acc)
        end)
      end
    end
  end

  @spec minimum_length_encoding(words :: [String.t()]) :: integer
  def minimum_length_encoding(words) do
    words
    |> Enum.reduce(Trie.new(), fn w, acc ->
      w |> String.graphemes() |> Enum.reverse() |> then(&Trie.insert(acc, &1))
    end)
    |> Trie.dfs(0, 0)
  end
end

File: algorithms/medium/short_encoding_of_words.rs
use std::collections::HashMap;

#[derive(Default)]
struct Trie {
    children: HashMap<char, Trie>,
}

impl Trie {
    fn new() -> Self {
        Self::default()
    }

    fn insert<I>(&mut self, iter: I) where I: Iterator<Item = char> {
        let mut curr = self;
        for c in iter {
            curr = curr.children.entry(c).or_default();
        }
    }

    fn dfs(&self, len: usize, sum: &mut usize) {
        let mut is_longest = true;
        for node in self.children.values() {
            is_longest = false;
            node.dfs(len + 1, sum);
        }
        if is_longest {
            *sum += len + 1;
        }
    }
}

impl Solution {
    pub fn minimum_length_encoding(words: Vec<String>) -> i32 {
        let mut tree = Trie::new();
        words.iter().for_each(|w| tree.insert(w.chars().rev()));
        let mut ans = 0;
        tree.dfs(0, &mut ans);
        ans as i32
    }
}
